numpy/core/memmap.py

```diff
@@ -309,3 +309,20 @@ def flush(self):
         """
         if self.base is not None and hasattr(self.base, 'flush'):
             self.base.flush()
+
+    def __array_wrap__(self, arr, context=None):
+        # Return memmap if self and arr are likely to share memory
+        if np.may_share_memory(self, arr):
```

Comment:
So, a remaining suggestion is to replace this statement with the simpler (and faster)
```
# Return the object it is was given as an output and should thus preserve type
if arr is self:
    return arr
```